File: development/lab04/lab04example03/src/lab04example03_node.cpp
#include <ros/ros.h>
#include <tf2_eigen/tf2_eigen.h>
#include <Eigen/Dense>
#include <tf2_ros/transform_broadcaster.h>
#include <tf2_kdl/tf2_kdl.h>
#include <sensor_msgs/JointState.h>
#include <kdl/chain.hpp>
#include <kdl/chainfksolverpos_recursive.hpp>






double a[6] = {0.0, 0.265699, 0.03, 0.0, 0.0, 0.0};
double alpha[6] = {-M_PI_2, 0.0, -M_PI_2, -M_PI_2, -M_PI_2, 0.0};
double d[6] = {0.159, 0.0, 0.0, 0.258, 0.0, -0.123};
double theta[6] = {0.0, -M_PI_2 + atan(0.03/0.264), -atan(0.03/0.264), 0.0, 0.0, 0.0};






KDL::Chain make_chain(void){
    KDL::Chain c;
    for (int i=0;i<6;i++){
        c.addSegment(KDL::Segment(KDL::Joint(KDL::Joint::RotZ), KDL::Frame::DH(a[i], alpha[i], d[i], theta[i])));
    }
    return c;
}


void fkine(const sensor_msgs::JointState::ConstPtr& joint_msg, tf2_ros::TransformBroadcaster br, KDL::Chain chain) {

    Eigen::Matrix4d A;
    Eigen::Matrix4d T = Eigen::Matrix4d::Identity(4, 4);
    KDL::ChainFkSolverPos_recursive FkSolver= KDL::ChainFkSolverPos_recursive(chain);

    geometry_msgs::TransformStamped transform[6];
    unsigned int nj = chain.getNrOfJoints();
    KDL::JntArray JointPositions = KDL::JntArray(nj);


    transform[0].header.frame_id = transform[1].header.frame_id = transform[2].header.frame_id =
    transform[3].header.frame_id = transform[4].header.frame_id = transform[5].header.frame_id = "world";

    transform[0].header.stamp = transform[1].header.stamp = transform[2].header.stamp =
    transform[3].header.stamp = transform[4].header.stamp = transform[5].header.stamp = ros::Time::now();

    transform[0].child_frame_id = "kdl_link_1";
    transform[1].child_frame_id = "kdl_link_2";
    transform[2].child_frame_id = "kdl_link_3";
    transform[3].child_frame_id = "kdl_link_4";
    transform[4].child_frame_id = "kdl_link_5";
    transform[5].child_frame_id = "kdl_link_6";

    for (int i = 0; i < nj; i++)
    {
        
        if ((i == 5) || (i == 4)){
            JointPositions(i) = -joint_msg->position.at(i);
        }
        else{
            JointPositions(i) = joint_msg->position.at(i);
        }
    }
    KDL::Frame FrameT;
    for (int i = 0; i < nj; i++)
    {
        FkSolver.JntToCart(JointPositions,FrameT, i+1);

        Eigen::Affine3d T_affine;

        T_affine.matrix() = T;

        geometry_msgs::TransformStamped T_buffer = tf2::kdlToTransform(FrameT);

        transform[i].transform = T_buffer.transform;

        br.sendTransform(transform[i]);
    }
}


int main(int argc, char **argv)
{
    ros::init(argc, argv, "robotis_fkine_node");
    ros::NodeHandle nh;
    tf2_ros::TransformBroadcaster br;
    KDL::Chain chain = make_chain();


    ros::Subscriber joint_sub_standard = nh.subscribe<sensor_msgs::JointState>("/joint_states", 10, boost::bind(fkine, _1, br, chain));

    ros::spin();
}

File: development/lab04/lab04example03/scripts/lab04example03_node.py
#!/usr/bin/env python

import rospy
from math import pi
import numpy as np
from sensor_msgs.msg import JointState
from geometry_msgs.msg import TransformStamped, Quaternion
import tf2_ros
import PyKDL as kdl


chain=kdl.Chain()



a = [0.0, 0.265699, 0.03, 0.0, 0.0, 0.0]
alpha = [-pi/2, 0.0, -pi/2, -pi/2, -pi/2, 0.0]
d = [0.159, 0.0, 0.0, 0.258, 0.0, -0.123]
theta = [0.0, -pi/2+np.arctan(0.03/0.264), -np.arctan(0.03/0.264), 0.0, 0.0, 0.0]
name_link = ['kdl_link_1', 'kdl_link_2', 'kdl_link_3', 'kdl_link_4', 'kdl_link_5', 'kdl_link_6']


chain=kdl.Chain()

for p_a, p_alpha, p_d, p_theta in zip(a, alpha, d, theta):
    chain.addSegment(kdl.Segment(kdl.Joint(kdl.Joint.RotZ), kdl.Frame().DH(a=p_a,alpha=p_alpha, d=p_d, theta=p_theta)))

fk=kdl.ChainFkSolverPos_recursive(chain)

##Declaring a transformBroadcaster like this is not a very optimal way.
##The very optimal way is to create a "class" for it.
##We will ask you to do this in the next Coursework. :)

br = tf2_ros.TransformBroadcaster()


def fkine_kdl(joint_msg):

    # convert joint_msg to a kdl joint array    
    joint_positions = kdl.JntArray(chain.getNrOfJoints())

    for i in range(chain.getNrOfJoints()):
        if (i==4 or i==5):
            joint_positions[i] = - joint_msg.position[i]
        else:
            joint_positions[i] = joint_msg.position[i]

    transform = TransformStamped()
    frame_t = kdl.Frame()

    for i in range(chain.getNrOfJoints()):
        transform.header.stamp = rospy.Time.now()
        transform.header.frame_id = 'world'
        transform.child_frame_id = name_link[i]
        #solve foraward kinematics up to link i
        fk.JntToCart(joint_positions, frame_t, i+1)

        transform = TransformStamped()
        transform.header.stamp = rospy.Time.now()
        transform.header.frame_id = 'world'
        transform.child_frame_id = name_link[i]

        transform.transform.translation.x = frame_t.p[0]
        transform.transform.translation.y = frame_t.p[1]
        transform.transform.translation.z = frame_t.p[2]
        transform.transform.rotation = Quaternion(*frame_t.M.GetQuaternion())

        br.sendTransform(transform)


def fkine_main():
    rospy.init_node('fkine_kdl_node')

    sub = rospy.Subscriber('/joint_states', JointState, fkine_kdl)
    
    rate = rospy.Rate(10)
    rospy.spin()
    rate.sleep()


if __name__ == '__main__':
    try:
        fkine_main()
    except rospy.ROSInterruptException:
        pass